New QL LED driver

spingaliQL opened the pull request:
I have added a new led driver under drivers/led. It uses FPGA IP, which **as of now** doesn't give any option to select LED color, blink duration, etc. 
FPGA bit stream is stored in a header file. The driver itself loads the FPGA, instead of manually loading outside through Jlink. The driver also does the IO mux setting and FBIO selection.
I have added a sample qf_helloworldhw under samples, which enable config for this driver.
I have added a new function in soc.c to select FBIOs. 
In soc.c:eos_s3_init() the function program_fpga_ip() is called based on a macro enable. If any driver wants to wants FPGA to be programmed it will define this function and enable this macro.
NOTE: as I'm merging all changes from Quicklogic-corp repo, the manifest file is also getting merged, which is modified to take hal_quicklogic repo from Quicklogic-corp. I'll undo this change on antmicro/zephyr after merging all these changes.

spingaliQL committed:
Modified e0s-s3-hal remote path

spingaliQL committed:
Merge pull request #1 from QuickLogic-Corp/manifest_changes

Modified eos-s3-hal remote path

spingaliQL committed:
added new led driver for S3. Added qf_helloworldhw sample. The led driver loads FPGA IP to blink an LED

spingaliQL committed:
Merge pull request #2 from QuickLogic-Corp/qf_helloworldhw

added new led driver for S3. Added qf_helloworldhw sample. The led dr…

spingaliQL committed:
Load litex PWM FPGA from M4

spingaliQL committed:
Merge pull request #3 from QuickLogic-Corp/eos_s3_pwm_loader

Load litex PWM FPGA from M4

spingaliQL committed:
moved qf_hellowworld sample to led driver samples

spingaliQL committed:
Merge pull request #4 from QuickLogic-Corp/eos_s3_led_sample1

moved qf_hellowworld sample to led driver samples

spingaliQL committed:
Fix:Removed Hello world print.

spingaliQL committed:
Merge pull request #5 from QuickLogic-Corp/eos_s3_led_sample1

Fix:Removed Hello world print.

spingaliQL committed:
Changed offset and added sleep after write

spingaliQL commented:
I have further made changes. I have moved "qf_helloworldhw"  from samples folder to samples/drivers folder and renamed as "led_eos_s3_1".
I have made changes to load litex PWM FPGA from M4 code. This driver is used in rgb_led and fade_led samples.

kowalewskijan commented on boards/arm/quick_feather/Kconfig.defconfig:
```diff
@@ -11,7 +11,7 @@ config BOARD
 
 config PWM_LITEX
-	select SOC_EOS_S3_FPGA
+	select EOS_S3_PROGRAM_FPGA
```
Please do not change this selection, because `SOC_EOS_S3_FPGA` configuration tells software if we want to initialize FPGA or not [here](https://github.com/antmicro/zephyr/blob/eos-s3-support/soc/arm/quicklogic_eos_s3/soc.c#L173). Without it clocks won't be set up.

spingaliQL replied:
The main change I made in these commits is to program FPGA from M4 program itself, instead of doing it outside using Jlink. To enable this feature I have added EOS_S3_PROGRAM_FPGA. If someone wants to use the old way, one can enable  SOC_EOS_S3_FPGA. I have not made any changes in this flow. Since all clocks are enabled with EOS_S3_PROGRAM_FPGA, other is not required. So EOS_S3_PROGRAM_FPGA and  SOC_EOS_S3_FPGA are mutually exclusive.

kowalewskijan replied:
Does `eos_s3_pwm_ip.h` contain `PWM_LITEX`? If not you must add separate config entry like explained [here](https://github.com/antmicro/zephyr/pull/3/files#r444115629).

spingaliQL replied:
No it doesn't.
I'll add separate config entry.

spingaliQL replied:
Added new configuration. However I understand from our H/W team that we are using litex pwm IP only, the bit stream is generated using symbi flow tools (with EOS S3 support). Our driver just loads the FPGA IP. It doesn't have full functions. For that I have to use pwm litex driver. So I have included that too.

kowalewskijan commented on drivers/led/CMakeLists.txt:
```diff
@@ -4,5 +4,6 @@ zephyr_sources_ifdef(CONFIG_HT16K33 ht16k33.c)
 zephyr_sources_ifdef(CONFIG_LP3943 lp3943.c)
 zephyr_sources_ifdef(CONFIG_LP5562 lp5562.c)
 zephyr_sources_ifdef(CONFIG_PCA9633 pca9633.c)
+zephyr_sources_ifdef(CONFIG_EOS_S3_LED1 led_eos_s3_1.c)
```
Can we just name it `led_eos_s3.c` ?

spingaliQL replied:
We are planning to have one more version of LED controller. I'll call this led_eos_s3_basic.c. "_1" is not giving good meaning. I'll apply this change in all related files/config.

kowalewskijan commented on drivers/led/Kconfig:
```diff
@@ -24,5 +24,6 @@ source "drivers/led/Kconfig.ht16k33"
 source "drivers/led/Kconfig.lp3943"
 source "drivers/led/Kconfig.lp5562"
 source "drivers/led/Kconfig.pca9633"
+source "drivers/led/Kconfig.eos_s3_1"
```
Can we just name it `Kconfig.eos_s3` ?

spingaliQL replied:
I'll rename to Kconfig.eos_s3_basic

kowalewskijan commented on drivers/led/Kconfig.eos_s3_1:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2018 Workaround GmbH
```
Change copyrights

spingaliQL replied:
ok

kowalewskijan commented on drivers/led/Kconfig.eos_s3_1:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2018 Workaround GmbH
+# SPDX-License-Identifier: Apache-2.0
+
+config EOS_S3_LED1
```
Without `1`? I'm not sure why you use this nomenclature with `1` at the end. Will it be more versions of driver?

spingaliQL replied:
We have a simple LED controller and an advanced one. I'll call all "1" as "basic"

kowalewskijan commented on drivers/led/eos_s3_led1_ip.h:
```diff
@@ -0,0 +1,1901 @@
+uint32_t	axFPGABitStream[] = {
```
Generally speaking storing a bitstream in a form of header in main Zephyr repository makes it unnecessary heavy. It will be much better to store it in an external repository - submodule like e.g. HAL and just initialize the submodule via `west`. I suggest to create a submodule for example `quicklogic_ip_library` and add it to `west` manifest.

spingaliQL replied:
We are planning to have a separate repo to store our IPs and drivers. I'm planning to bring that repo as a sub module under "eos-s3-hal" repo, which is already defined in zephyr manifest file. 
I see `west update` doesn't clone the module repo recursively. I think after `west update`, we have to do `git submodule init` and `git submodule update`. Please suggest if there is a better way.

spingaliQL replied:
> We are planning to have a separate repo to store our IPs and drivers. I'm planning to bring that repo as a sub module under "eos-s3-hal" repo, which is already defined in zephyr manifest file.
> I see `west update` doesn't clone the module repo recursively. I think after `west update`, we have to do `git submodule init` and `git submodule update`. Please suggest if there is a better way.

Do you have any suggestion on this point?

kowalewskijan replied:
I suppose it didn't clone because there was no `zephyr/module.yml` file in repository with IPs and/or there was no entry in `west.yml` in main zephyr repo. If you want to add IP repository as submodule to `hal_quicklogic` please do as follow:
* Add to `west.yml` in your main zephyr repository entry:
```
    - name: <name_of_module>
      remote: QuickLogic-Corp
      repo-path: <repo_name>
      revision: <SHA>
      path: modules/hal/quicklogic/<name_of_module>
```
* Add proper remote in `remotes` section of `west.yml`:
```
    - name: QuickLogic-Corp
      url-base: https://github.com/QuickLogic-Corp
```
* Add proper `CMakeLists.txt` and `zephyr/module.yml` to repository you want to add (take a look at these from [hal_quicklogic](https://github.com/antmicro/hal_quicklogic/tree/initial_release))
* Run `west update`

I couldn't find any examples of storing bitstreams in Zephyr mainline, so I suppose this could be a good start for now. Anyway you can always extract submodule with IPs from `hal_quicklogic` and make it an independent Zephyr module in future.

kowalewskijan commented on drivers/led/led_eos_s3_1.c:
```diff
@@ -0,0 +1,49 @@
+
+void program_fpga_ip()
+{
+	// Load bitstrem into FPGA
```
Please use `/* */` to comment code

spingaliQL replied:
ok. Resolved all coding style problems using the script you suggested. Thank you.

kowalewskijan commented on drivers/led/led_eos_s3_1.c:
```diff
@@ -0,0 +1,49 @@
+void program_fpga_ip()
+{
+	// Load bitstrem into FPGA
+	load_fpga(sizeof(axFPGABitStream),axFPGABitStream);
```
Missing `space`:
```suggestion
	load_fpga(sizeof(axFPGABitStream), axFPGABitStream);
```

kowalewskijan commented on drivers/led/led_eos_s3_1.c:
```diff
@@ -0,0 +1,49 @@
+	// Load bitstrem into FPGA
+	load_fpga(sizeof(axFPGABitStream),axFPGABitStream);
+
+	// Configure IOs
```
`/* */`

kowalewskijan commented on drivers/pwm/CMakeLists.txt:
```diff
@@ -15,6 +15,7 @@ zephyr_library_sources_ifdef(CONFIG_PWM_SAM		pwm_sam.c)
 zephyr_library_sources_ifdef(CONFIG_PWM_MCUX		pwm_mcux.c)
 zephyr_library_sources_ifdef(CONFIG_PWM_XEC		pwm_mchp_xec.c)
 zephyr_library_sources_ifdef(CONFIG_PWM_LITEX		pwm_litex.c)
+zephyr_library_sources_ifdef(CONFIG_PWM_LITEX		pwm_eos_s3.c)
```
Change to `CONFIG_PWM_EOS_S3`

spingaliQL replied:
ok. Will do.

kowalewskijan commented on drivers/pwm/pwm_eos_s3.c:
```diff
@@ -0,0 +1,33 @@
+
+void program_fpga_ip()
+{
+	// Load bitstrem into FPGA
```
`/* */`

kowalewskijan commented on samples/drivers/led_eos_s3_1/src/main.c:
```diff
@@ -0,0 +1,23 @@
+    	printk( __DATE__ " " __TIME__ "\n" );
+    	printk( "##########################\n\n");
+
+	while (1);
```
I think it is not needed since controller in FPGA will drive LED infinitely, is that correct? When Zephyr finishes `main` function it switches to `idle` mode, so there is no reason for busy wait.

spingaliQL replied:
Correct. Removed the line.

kowalewskijan commented on boards/arm/quick_feather/Kconfig.defconfig:
```diff
@@ -11,7 +11,7 @@ config BOARD
-	select SOC_EOS_S3_FPGA
+	select EOS_S3_PROGRAM_FPGA
 	default y
 
```
If you want to use your own driver instead of `PWM_LITEX` please add proper configs here like:
```suggestion
config PWM_EOS_S3
    select SOC_EOS_S3_FPGA
    select EOS_S3_PROGRAM_FPGA
    default y
```
and then disable `PWM_LITEX` in `prj.conf` in sample: `CONFIG_PWM_LITEX=n`

spingaliQL replied:
Earlier the FPGA was loaded through JLink. As part of eos_s3_init() FPGA clocks used to get initialized. Instead of this now the FPGA bit stream is loaded to EOS S3. Clock init happens as part of it. 
In both cases we have been using our FPGA IP. There is no change in the flow. That's why I have not created a new config. If you still thing I have to go with new config, I'll do. Please let me know.

kowalewskijan commented on drivers/pwm/pwm_eos_s3.c:
```diff
@@ -0,0 +1,33 @@
+void program_fpga_ip()
+{
+	// Load bitstrem into FPGA
+	load_fpga(sizeof(axFPGABitStream),axFPGABitStream);
```
```suggestion
	load_fpga(sizeof(axFPGABitStream), axFPGABitStream);
```

kowalewskijan commented on soc/arm/quicklogic_eos_s3/CMakeLists.txt:
```diff
@@ -5,4 +5,5 @@ zephyr_include_directories(${ZEPHYR_BASE}/drivers)
 zephyr_sources(
   soc.c
   irq_handlers.c
+  fpga_loader.c
```
`fpga_loader.c` should be added to compilation only when we want to use FPGA.
```suggestion
```
Please add `zephyr_sources_ifdef(CONFIG_EOS_S3_PROGRAM_FPGA fpga_loader.c` below

spingaliQL replied:
Done. Looks neat with this change. Thank you.

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+#define CFG_CTL_CFG_DATA                (*(volatile uint32_t *)(0x40014FFC))
+#define CFG_CTL_CFG_CTL                 (*(volatile uint32_t *)(0x40014000))
+
+static void enable_clocks()
```
Could you compare your function to this one that is already implemented in `soc.c`? ([link](https://github.com/antmicro/zephyr/blob/eos-s3-support/soc/arm/quicklogic_eos_s3/soc.c#L115-L158)). I think we can use the function from `soc.c` here instead of writing a new one.

spingaliQL replied:
Yes. There is a duplication. enable_clocks() is a subset of eos_s3_fpga_init(). I can not use eos_s3_fpga_init as other initializations are not required . Specially PMU->GEN_PURPOSE_0=0 will not allow FPGA programming. 
I'll rename enable_clocks() as enable_fpga_clocks() and move it to soc.c. I'll call this function in both places instead of duplicating.

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+		PMU->GEN_PURPOSE_1  = i << 4;
+	}
+
+	REG8 = 0x10;
```
Can we make an array e.g. `initialization_sequence` and write to `REG8` in a loop?

spingaliQL replied:
This is for adding delay. I tried to use k_sleep. Somehow it is not working. System hangs at k_sleep. I guess we can not use k_sleep at this place, since we are still in system init function. Is is correct?

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+
+	enable_clocks();
+
+	// Configuration of CFG_CTRL for writes
```
Change comment by using `/* */`

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+
+	// Configuration of CFG_CTRL for writes
+	CFG_CTL_CFG_CTL = 0x0000bdff ;
+	// wait some time for fpga to get reset pulse
```
Change comment by using `/* */`

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+	REG8 = 0x70;
+	REG8 = 0x80;
+
+	REG9 = 0xBDFF;
```
There are a plenty of `magic` values, could you check if these values are defined somewhere in HAL? It would greatly improve the readability.

spingaliQL replied:
ok. I'll find out.

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+	for(chunk_cnt=0;chunk_cnt<(img_size/4);chunk_cnt++)
+		CFG_CTL_CFG_DATA = gFPGAPtr[chunk_cnt];
+
+	// wait some time for fpga to get reset pulse
```
`/* */`

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+		PMU->GEN_PURPOSE_1  = i << 4;
+	}
+
+	CFG_CTL_CFG_CTL = 0x0; // exit config mode
```
`/* */`

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+	REG14 = 0x90;
+
+
+	PMU->GEN_PURPOSE_0 = 0x0; //set APB_FB_EN = 0 for normal mode
```
`/* */`

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+
+	PMU->GEN_PURPOSE_0 = 0x0; //set APB_FB_EN = 0 for normal mode
+
+	// required wait time before releasing LTH_ENB
```
`/* */`

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+	PMU->GEN_PURPOSE_0 = 0x0; //set APB_FB_EN = 0 for normal mode
+
+	// required wait time before releasing LTH_ENB
+	for (i=0;i<500; i++) {
```
```suggestion
	for (i = 0; i < 500; i++) {
```

spingaliQL replied:
Fixed all styling problems using script/manual. Marking all of these resolved. Thank you.

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+
+	// required wait time before releasing LTH_ENB
+	for (i=0;i<500; i++) {
+		PMU->GEN_PURPOSE_1  = i << 4;
```
This loop is just for waiting? Or it does something more? If you want to wait for a certain time you can use `k_sleep(time_in_miliseconds)`

spingaliQL replied:
I'll use k_sleep

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+		PMU->GEN_PURPOSE_1  = i << 4;
+	}
+
+	//release isolation - LTH_ENB
```
`/* */`

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+	PMU->FB_ISOLATION = 0x0;
+	*((volatile unsigned int*) 0x40004c4c) = 0x000009a0;
+
+	printf("FPGA is programmed\r\n");
```
If it is a debug message, please remove it.

spingaliQL replied:
Agree. Removed.

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,153 @@
+
+	printf("FPGA is programmed\r\n");
+
+	return 1;
```
This is confusing, because if it is a successful return we should return 0
```suggestion
	return 0;
```

spingaliQL replied:
This happens when someone moves from windows to linux!
Changed as suggested.

kowalewskijan commented on soc/arm/quicklogic_eos_s3/soc.c:
```diff
@@ -169,6 +193,10 @@ static int eos_s3_init(struct device *arg)
 	eos_s3_lock_disable();
 
+#ifdef CONFIG_EOS_S3_PROGRAM_FPGA
+	program_fpga_ip();
```
I would move it after `eos_s3_fpga_init();`

spingaliQL replied:
Programming FPGA comes first in the sequence of operation. When it is done manually outside, we would have done it before entering inside this function.

kowalewskijan commented on soc/arm/quicklogic_eos_s3/soc.h:
```diff
@@ -65,6 +66,8 @@ void EOSS3_DisableIRQ(IRQn_Type IRQn);
 void EOSS3_EnableIRQ(IRQn_Type IRQn);
 void EOSS3_ClearPendingIRQ(IRQn_Type IRQn);
 
+void program_fpga_ip();
```
```suggestion
void program_fpga_ip(void);
```

kowalewskijan commented on drivers/led/led_eos_s3_1.c:
```diff
@@ -0,0 +1,49 @@
+	
+}
+
+void program_fpga_ip()
```
```suggestion
void program_fpga_ip(void)
```

kowalewskijan commented on drivers/pwm/pwm_eos_s3.c:
```diff
@@ -0,0 +1,33 @@
+#include <fpga_loader.h>
+#include "eos_s3_pwm_ip.h"
+
+void program_fpga_ip()
```
```suggestion
void program_fpga_ip(void)
```

spingaliQL replied:
Using the script I resolved these style problems. Thank you.

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.h:
```diff
@@ -0,0 +1,23 @@
+*                                                          
+*=========================================================*/
+
+extern int load_fpga(uint32_t img_size,uint32_t* image_ptr);
```
Do we need `extern` here?

spingaliQL replied:
Removed

kowalewskijan commented:
@spingaliQL If you want to check your code formatting to fit Zephyr standards, you can use a script provided in Zephyr repository. It is stored in: `zephyr/scripts/checkpatch.pl`. To use it you just simply: `./checkpatch.pl -f path/to/file`

spingaliQL commented:
@kowalewskijan Thank you very much for detailed review and suggesting the script for code formatting. It is very helpful. I have started working on the changes recommended. Please see my replies and let me know if you have further comments.  Once I complete the changes, I'll inform you in same PR.

spingaliQL committed:
Changes for flash samples

-Change QF dts file with new partitions
-Removed sleep which was added for testing
-Reinitialize C02 clock divider so that it will not change after running
boot loader

spingaliQL committed:
Merge branch 'eos-s3-support' of https://github.com/QuickLogic-Corp/zephyr into code_review_changes1

spingaliQL committed:
Changes for Antmicro code review

-Added new config EOS_S3_PWM which loads litex pwm ip bit stream and then uses litex pwm driver
-Renamed EOS_S3_LED1 as EOS_S3_LED_BASIC; in configuration files, driver files and sample files
-Cleanup in fpga_loader.c file. Instead of using hardcoded register addresses using HAL defined macros
-Moved the common code used in soc.c and fpga_loader.c to a common function enable_fpga_clocks()
-Resolved coding style related issues in all .c files

spingaliQL committed:
Merge branch 'code_review_changes1' into spi_flash_samples

spingaliQL committed:
Merge pull request #6 from QuickLogic-Corp/code_review_changes1

Antmicro Code review changes

spingaliQL committed:
Merge branch 'eos-s3-support' of https://github.com/QuickLogic-Corp/zephyr into spi_flash_samples

spingaliQL committed:
Merge pull request #7 from QuickLogic-Corp/spi_flash_samples

Spi flash samples

spingaliQL reviewed (commented):
I have made below changes as per the review:
-Added new config EOS_S3_PWM which loads litex pwm ip bit stream and then uses litex pwm driver.
-Renamed EOS_S3_LED1 as EOS_S3_LED_BASIC; in configuration files, driver files and sample files 
-Cleanup in fpga_loader.c file. Instead of using hardcoded register addresses using HAL defined macros
-Moved the common code used in soc.c and fpga_loader.c to a common function enable_fpga_clocks()
-Resolved coding style related issues in all .c files

spingaliQL commented:
Hi @kowalewskijan 
I have resolved almost all review comments, except these:
-Moving the IP bitstream header files to separate repo: Our H/W team is working on the repo for keeping all our IPs. Once it is ready, I'll refer the header files from there and remove from drivers folder.
-I tried to use k_sleep in fpga_loader.c as you suggested. It doesn't work, system hangs. I don't know if k_sleep can be used here as we are still executing system init function. Please suggest how to use sleep function.
Thank you.

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,92 @@
+#if 0
+	k_sleep(100);
+#else
+	for (i = 0; i < 60; i++)
```
If we wait for reset pulse as stated above:
> /* wait some time for fpga to get reset pulse */

Maybe there is a kind of flag that could provide an information about it? If yes, then we could have just something like this:
```suggestion
	while (!REG->FLAG) {
		PMU->GEN_PURPOSE_1  = i << 4;
              k_sleep(1);
      }
```

kowalewskijan commented on soc/arm/quicklogic_eos_s3/soc.c:
```diff
@@ -109,28 +150,17 @@ static void eos_s3_cru_init(void)
+	/* Set C02 clock to default value,
+	 * if any prev app such as boot loader changes it will be set back
+	 */
+	CRU->CLK_CTRL_B_0 = 0x204;
```
Are there any macros in HAL that we can use instead of a magic value?

kowalewskijan commented on drivers/led/led_eos_s3_basic.c:
```diff
@@ -0,0 +1,49 @@
+
+}
+
+void program_fpga_ip(void)
```
I thought about loading bitstream inside other drivers and I came to conclusion that the best solution would be to introduce a new type of driver in Zephyr - FPGA manager. We should add `drivers/fpga` directory in Zephyr, add global `Kconfig` and `Kconfig.eos_s3` and add there a new driver exclusively for FPGA control, like: FPGA init, FPGA bitstream loading etc. So in the end there would be 2 separate layers - 1st: FPGA control related and 2nd: driver for IPcore. This would allow us to upload bitstream at early boot stage and later use IP core drivers without special FPGA managing functions.

kowalewskijan commented on samples/drivers/spi_flash/src/main.c:
```diff
@@ -21,7 +21,7 @@
 #endif
 
-#define FLASH_TEST_REGION_OFFSET 0xff000
+#define FLASH_TEST_REGION_OFFSET 0x100000
```
Please undo if this is a leftover

spingaliQL replied:
We have M4 application image starting from offset 0x0008’0000 to 0x000F’FFFF. So we can not use the location 0xff000 for this test. Since the offset is hard-coded within the application I have modified it for testing. Shall I change it to DT_FLASH_AREA_STORAGE_OFFSET as in nvs sample? If not, I don't know what is the way out.

Note: I have given image-3 size wrong as 0x20000 in quick_feather.dts. I'll fix it in next commit.

kowalewskijan replied:
I think we can fix it like [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/drivers/spi_flash/src/main.c#L29-L30)

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,92 @@
+	u32_t        chunk_cnt = 0;
+	volatile u32_t   *gFPGAPtr = (volatile u32_t *)image_ptr;
+
+	IO_MUX->PAD_19_CTRL = 0x00000180;
```
Are there any macros to use instead of a magic value?

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,92 @@
+	k_sleep(100);
+#else
+	for (i = 0; i < 60; i++)
+		PMU->GEN_PURPOSE_1  = i << 4;
```
There are multiple of these `for` loops, please add a static function dedicated to wait process

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,92 @@
+
+	CRU->FB_SW_RESET = 0;
+	CRU->FB_MISC_SW_RST_CTL = 0;
+	PMU->GEN_PURPOSE_1 = 0x90;
```
Are there any macros to use instead of a magic value?

kowalewskijan commented on soc/arm/quicklogic_eos_s3/fpga_loader.c:
```diff
@@ -0,0 +1,92 @@
+	for (i = 0; i < 60; i++)
+		PMU->GEN_PURPOSE_1  = i << 4;
+#endif
+	IO_MUX->PAD_19_CTRL = 0x000009a0;
```
Are there any macros to use instead of a magic value?

kowalewskijan commented on soc/arm/quicklogic_eos_s3/soc.c:
```diff
@@ -143,6 +173,7 @@ static void eos_s3_fpga_init(void)
 
 	CRU->C08_X4_CLK_GATE = C08_X4_CLK_GATE_PATH_0_ON;
 
+
```
remove redundant line

spingaliQL committed:
spi flash sample offset fix

Kept the original offset as it is. For quick feather board offset is
changed to 0x100000
Corrected the image-3 size in quick feather dts file

spingaliQL committed:
Merge pull request #8 from QuickLogic-Corp/spi_flash_sample_fix

spi flash sample offset fix
